cpu/cortexm_common/periph/systick.c

```diff
@@ -0,0 +1,190 @@
+ * @brief       Low-level timer driver implementation based on SysTick
+ *
+ *              SysTick is a 24-bit Down-Counting Timer
+ *              It runs at the same frequency as the CPU
+ *              The count (VAL) register can only be set to 0 on write
+ *
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ * @}
+ */
+
+#include <assert.h>
+#include "board.h"
+#include "periph_conf.h"
+#include "systick.h"
+
+/* SysTick CTRL configuration */
+#define ENABLE_MASK     (SysTick_CTRL_CLKSOURCE_Msk |\
+                         SysTick_CTRL_ENABLE_Msk    |\
+                         SysTick_CTRL_TICKINT_Msk)
+
+#define ALARM_EXT_MAX   ((1UL << (8 * sizeof(alarm_ext))) - 1)
+
+/* interrupt is always enabled to update time */
+static bool alarm_enabled;
+
+/* divider for CLOCK_CORECLOCK */
+static uint16_t prescaler = 1;
+
+/* full periods before alarm */
+static uint16_t alarm_ext;
+
+/* remaining time to alarm (last period length) */
+static uint32_t alarm_val;
+
+/* current time if the timer ran at freq Hz */
+static uint32_t now;
+
+/* offset in CLOCK_CORECLOCK ticks */
+static uint32_t now_offset;
+
+/* dummy SysTick callback */
+__attribute__((weak))
+void systick_callback(void)
+{
+    /* no-op */
+}
+
+void systick_timer_init(unsigned long freq)
+{
+    if (IS_USED(MODULE_SYSTICK_PRESCALER)) {
+        assert(freq <= CLOCK_CORECLOCK);
+
+        prescaler = CLOCK_CORECLOCK / freq;
+        assert(CLOCK_CORECLOCK % freq == 0);
+    } else {
+        assert(freq == CLOCK_CORECLOCK);
+    }
+
+    SysTick->VAL  = 0;
+    SysTick->LOAD = SysTick_LOAD_RELOAD_Msk;
+    SysTick->CTRL = ENABLE_MASK;
+
+    NVIC_SetPriority(SysTick_IRQn, CPU_CORTEXM_PENDSV_IRQ_PRIO);
+    NVIC_EnableIRQ(SysTick_IRQn);
+}
+
+int systick_timer_set(unsigned int value)
+{
+    uint64_t val64 = value * prescaler;
+
+    /* check if value fits with prescaler */
+    if ((val64 >> 24) > ALARM_EXT_MAX) {
+        return -1;
+    }
+
+    /* stop timer to prevent race condition */
+    SysTick->CTRL = 0;
+
+    alarm_ext = val64 >> 24;
+    alarm_val = val64 & SysTick_LOAD_RELOAD_Msk;
+
+    if (IS_USED(MODULE_SYSTICK_PRESCALER)) {
+        now_offset += SysTick->LOAD - SysTick->VAL;
+    } else {
+        now += SysTick->LOAD - SysTick->VAL;
+    }
+
+    if (alarm_ext) {
+        SysTick->LOAD = SysTick_LOAD_RELOAD_Msk;
+    } else {
+        SysTick->LOAD = alarm_val;
+    }
+
+    /* force reload */
+    SysTick->VAL = 0;
+
+    /* start the timer again */
+    SysTick->CTRL = ENABLE_MASK;
+
+    /* check if the last period is after the next reload */
+    if (alarm_ext == 1) {
+        SysTick->LOAD = alarm_val;
+    }
+
+    alarm_enabled = 1;
+
+    return 0;
+}
+
+int systick_timer_set_absolute(unsigned int value)
+{
+    return systick_timer_set(value - systick_timer_read());
+}
+
+void systick_timer_clear(void)
+{
+    SysTick->LOAD = SysTick_LOAD_RELOAD_Msk;
+    alarm_enabled = 0;
+}
+
+unsigned int systick_timer_read(void)
+{
```

Comment:
I don't think this safe regarding disabled interrupts (or if called within an isr).

1. (LOAD - VAL) is almost at LOAD
2. isrs get disabled
3. VAL overflows (reloads to LOAD), but systick isr cannot update "now"
4. this function misses a period

Reply:
2768ac7 should address that.